src/isolation/landlock.rs

```diff
@@ -0,0 +1,356 @@
+		// whitelisted files themselves, but apply to the contents of directories transitively.
+		// This means that we have to give Uhyve read-only access to the parent directory with
+		// some additional permissions so as to support fs::remove_file and removing the temporary
+		// directory.
+		let access_dir_with_rm: landlock::BitFlags<AccessFs, u64> =
+			AccessFs::ReadDir | AccessFs::RemoveDir | AccessFs::RemoveFile;
+
+		let res_status = Ruleset::default()
+			.handle_access(access_all)?
+			.create()?
+			.add_rules(
+				self.rw_paths
+					.as_slice()
+					.iter()
+					.map::<Result<_, RestrictError>, _>(|p| {
+						debug!("Adding read-write path ruleset for {:#?}", *p);
+						Self::determine_ruleset(PathFd::new(p)?, abi)
+					}),
+			)?
+			.set_compatibility(self.compat_level)
+			.add_rules(
+				self.ro_paths
+					.as_slice()
+					.iter()
+					.map::<Result<_, RestrictError>, _>(|p| {
+						debug!("Adding read-only path ruleset for {:#?}", *p);
+						Self::determine_ruleset(PathFd::new(p)?, abi)
+					}),
+			)?
+			.set_compatibility(self.compat_level)
+			.add_rules(
+				self.ro_dirs
+					.as_slice()
+					.iter()
+					.map::<Result<_, RestrictError>, _>(|p| {
+						debug!("Adding read-only directory ruleset for {:#?}", *p);
+						Ok(PathBeneath::new(PathFd::new(p)?, access_dir_with_rm))
+					}),
+			)?
+			.set_compatibility(self.compat_level)
+			.set_no_new_privs(true)
+			.restrict_self()?;
+
+		debug!(
+			"Landlock ruleset status: {:#?} (no_new_privs: {:#?})",
+			res_status.ruleset, res_status.no_new_privs
+		);
+
+		// Adapted from Mickaël Salaün's Rust Landlock library, which is distributed
+		// under the terms of the MIT and Apache 2.0 licenses.
+		//
+		// For the original source code authored by Mickaël Salaün, see:
+		// https://github.com/landlock-lsm/rust-landlock/blob/07f2a9a7/examples/sandboxer.rs#L133C1-L135C6
+		if res_status.ruleset == RulesetStatus::NotEnforced {
+			panic!(
+				"Landlock is not supported by the running kernel. You can disable Landlock using `--file-isolation none`."
+			);
+		}
+
+		let mut is_already_enabled = LANDLOCK_ENABLED.lock().unwrap();
+		*is_already_enabled = true;
+
+		Ok(res_status)
+	}
+
+	/// Derive a suitable Landlock policy for a given file descriptor.
+	///
+	/// Although applying directory-specific access rights to a file will just result in those rights being ignored
+	/// in "best effort mode", this is not the case when running Uhyve
+	/// in strict mode. Therefore, this function omits directory-specific
+	/// access rights to "R/W" files.
+	///
+	/// - `fd` - File descriptor (of type [`PathFd`])
+	/// - `abi` - Landlock ABI version
+	fn determine_ruleset(fd: PathFd, abi: ABI) -> Result<PathBeneath<PathFd>, RestrictError> {
+		let is_file = Self::is_file(fd.as_fd().as_raw_fd()).unwrap();
+		match is_file {
+			true => Ok(PathBeneath::new(fd, AccessFs::from_file(abi))),
+			false => Ok(PathBeneath::new(fd, AccessFs::from_all(abi))),
+		}
```

Comment:
```suggestion
		OK (
                if Self::is_file(fd.as_fd().as_raw_fd()).unwrap() {
		                  PathBeneath::new(fd, AccessFs::from_file(abi))
                } else {
		                  PathBeneath::new(fd, AccessFs::from_all(abi))
                }
        )
```

Can this function actually return an `Err`?